compiler/rustc_parse/src/validate_attr.rs

```diff
@@ -167,7 +167,13 @@ pub fn check_attribute_safety(psess: &ParseSess, safety: AttributeSafety, attr:
             // square bracket respectively.
             let diag_span = attr_item.span();
 
-            if attr.span.at_least_rust_2024() {
+            // Attributes can be safe in earlier editions, and become unsafe in later ones.
+            let emit_error = match unsafe_since {
+                None => true,
+                Some(unsafe_since) => attr.span.edition() >= unsafe_since,
+            };
```

Comment:
This is equivalent to `Some(attr.span.edition) >= unsafe_since` but that is very subtle so I wrote out the logic here. In general I don't love using `Option` here, but a custom enum would add a bunch of extra code and comes with its own naming problems.